src/ewasm/expressions/mod.rs

```diff
@@ -0,0 +1,1314 @@
+use crate::ewasm::function_context::FunctionContext;
+use crate::ewasm::statements::LLVMAssertion;
+use crate::ewasm::types::{get_type_as_string, LLVMType};
+use crate::ewasm::utils::*;
+use inkwell::values::PointerValue;
+use inkwell::AddressSpace;
+
+pub struct LLVMExpression<'a> {
+    pub expression: &'a Expression,
+}
+
+impl<'a> LLVMExpression<'a> {
+    // We want to take an expression, create any of the intermediary steps to evaluate it,
+    // and then return the variable that stores the evaluated result
+    pub fn generate<'ctx>(
+        &self,
+        codegen: &mut Codegen<'_, 'ctx>,
+        function_context: &mut FunctionContext<'ctx>,
+    ) -> Option<BasicValueEnum<'ctx>> {
+        match self.expression {
+            Expression::Identifier(i) => {
+                LLVMIdentifier { identifier: i }.generate(codegen, function_context)
+            }
+            Expression::BinaryExpression(b) => {
+                LLVMBinaryExpression { expression: b }.generate(codegen, function_context)
+            }
+            Expression::InoutExpression(i) => {
+                LLVMInoutExpression { expression: i }.generate(codegen, function_context)
+            }
+            Expression::ExternalCall(f) => {
+                LLVMExternalCall { external_call: f }.generate(codegen, function_context)
+            }
+            Expression::FunctionCall(f) => {
+                LLVMFunctionCall { function_call: f }.generate(codegen, function_context)
+            }
+            Expression::VariableDeclaration(v) => {
+                LLVMVariableDeclaration { declaration: v }.generate(codegen, function_context)
+            }
+            Expression::BracketedExpression(b) => LLVMExpression {
+                expression: &*b.expression,
+            }
+            .generate(codegen, function_context),
+            Expression::AttemptExpression(_) => panic!("Should be removed in the preprocessor"),
+            Expression::Literal(l) => {
+                LLVMLiteral { literal: l }.generate(codegen, function_context)
+            }
+
+            Expression::ArrayLiteral(a) => {
+                let elements = a
+                    .elements
+                    .iter()
+                    .map(|e| {
+                        LLVMExpression { expression: e }
+                            .generate(codegen, function_context)
+                            .unwrap()
+                    })
+                    .collect::<Vec<BasicValueEnum>>();
+
+                // If we have an empty list, we default to an empty int32 array
+                let elem_type = elements
+                    .first()
+                    .map_or(codegen.context.i32_type().as_basic_type_enum(), |val| {
+                        val.get_type()
+                    });
+                let arr_ptr = codegen
+                    .builder
+                    .build_alloca(elem_type.array_type(elements.len() as u32), "arr_ptr");
```

Comment:
Concerned with magic strings throughout this file, declare constants

Reply:
These strings do not do anything. They are just what the outputted LLVM ends up calling variables. We do not need to store or keep track of the names, they just aid readability of the LLVM output. My point is basically that an example such as this is not an example of fragile code where one might need to (or even be able to) try to look up a value by the name "arr_ptr" only to find it had been changed somewhere else.

That said of course, if you would still prefer they be made constants somewhere, it will be done

Reply:
You are choosing the names, for LLVM, correct? Do you rely on the names being as such anywhere?

Reply:
No, they are only "used" when dumped to the output llvm. LLVM does not even use them. We create something here called "arr_ptr" every time we need to point into an array. This may happen lots of times in the same scope, so LLVM automatically mangles it anyway, so it may not even end up as "arr_ptr"

Reply:
Sure, this is fine then